internal/handlers/common/update.go

```diff
@@ -236,6 +243,71 @@ func processPopFieldExpression(doc *types.Document, update *types.Document) (boo
 	return changed, nil
 }
 
+// processRenameFieldExpression changes document according to $rename operator.
+// If the document was changed it returns true.
+func processRenameFieldExpression(doc *types.Document, update *types.Document) (bool, error) {
+	renameExpression := update.SortFieldsByKey()
+
+	var changed bool
+
+	skipErrReg := regexp.MustCompile(`^types\.getByPath: types\.Document\.Get: key not found:.+`)
+
+	for _, key := range renameExpression.Keys() {
+		renameRawValue, err := renameExpression.Get(key)
+		if err != nil {
+			// if $rename field does not exist, don't change anything
+			continue
+		}
+
+		if key == "" || renameRawValue == "" {
+			return changed, NewWriteErrorMsg(ErrEmptyName, "An empty update path is not valid.")
+		}
+
+		sourcePath := types.NewPathFromString(key)
+
+		renameValue, ok := renameRawValue.(string)
+		if !ok {
+			return changed, NewWriteErrorMsg(
+				ErrBadValue,
+				fmt.Sprintf("the 'to' field for $rename must be a string: %s: %v", key, renameRawValue),
+			)
+		}
+
+		if key == renameRawValue {
+			return changed, NewWriteErrorMsg(
+				ErrBadValue,
+				fmt.Sprintf("The source and target field for $rename must differ: %s: %#v", key, renameValue),
+			)
+		}
+
+		targetPath := types.NewPathFromString(renameValue)
+
+		// Get value to move
+		val, err := doc.GetByPath(sourcePath)
+		if err != nil {
+			// TODO refactor `getByPath` to make this check more elegant
+			if skipErrReg.MatchString(err.Error()) {
+				continue
+			}
+
+			return changed, NewWriteErrorMsg(ErrUnsuitableValueType, err.Error())
+		}
+
+		// Remove old document
+		doc.RemoveByPath(sourcePath)
+
+		// Set new path with old value
+		err = doc.SetByPath(targetPath, val)
+		if err != nil {
+			return changed, err
+		}
+
+		changed = true
```

Comment:
Could we have a testcase with multiple rename fields with the first one succeeds and second one fails. Just want check we behave the same as mongoDB on setting `changed`

`bson.D{{"$rename", bson.D{{"v.foo", "boo"}}, bson.D{{"v.array", 1}}}}`
It will use dataset from here https://github.com/FerretDB/FerretDB/blob/main/integration/shareddata/composites.go#L33

Reply:
done

Reply:
Have I misunderstood something? :)
This is covered in [MultipleSecondInvalid](https://github.com/FerretDB/FerretDB/pull/1753/files#diff-02beb83969c6feb8b48d15aebcd299647e0aa2e8b1e5c531e74eb2c4f636ad85R537-R540)

Reply:
Yes this is covered now. Thanks for adding it!